pkg/status/combinedstatus-resolution.go

```diff
@@ -580,73 +565,180 @@ func handleSelect(scName string, scData *statusCollectorData) *v1alpha1.NamedSta
+	}
+
+	valValue := val.Value()
+	// temporary until type checking is implemented
+	switch v := valValue.(type) {
+	case string:
+		return v1alpha1.Value{
+			Type:   v1alpha1.TypeString,
+			String: &v,
+		}
+	case int, int8, int16, int32, int64:
+		numStr := strconv.Itoa(int(v.(int64)))
+		return v1alpha1.Value{
+			Type:   v1alpha1.TypeNumber,
+			Number: &numStr,
+		}
+	case uint, uint8, uint16, uint32, uint64:
+		numStr := strconv.FormatUint(v.(uint64), 10)
+		return v1alpha1.Value{
+			Type:   v1alpha1.TypeNumber,
+			Number: &numStr,
+		}
+	case float32, float64:
+		numStr := strconv.FormatFloat(v.(float64), 'g', -1, 64)
+		return v1alpha1.Value{
+			Type:   v1alpha1.TypeNumber,
+			Number: &numStr,
+		}
+	case bool:
+		return v1alpha1.Value{
+			Type: v1alpha1.TypeBool,
+			Bool: &v,
+		}
+	default:
+		evalJSON, err := json.Marshal(valValue)
+		if err != nil {
+			runtime2.HandleError(fmt.Errorf("failed to marshal select evaluation: %w", err))
+			return v1alpha1.Value{
+				Type: v1alpha1.TypeNull,
+			}
+		} else {
+			return v1alpha1.Value{
+				Type:   v1alpha1.TypeObject,
+				Object: &extv1.JSON{Raw: evalJSON},
+			}
 		}
-	} else {
-		// if there is no groupBy, create a single group
-		groupByNameToValueToWorkStatus[allValue] = make(map[any][]*workStatusData)
 	}
+}
 
-	// fill the inner layer
-	for _, wsData := range scData.wecToData {
-		for _, groupByNamedExp := range scData.GroupBy {
-			groupByValue := wsData.groupByEval[groupByNamedExp.Name]
-			if groupByValue == nil {
-				continue
+// handleAggregation handles the aggregation of the statuscollector data.
+// This means that the function evaluates the groupBy expressions and the
+// combinedFields expressions against the possibly filtered workstatuses and
+// returns the result in a NamedStatusCombination.
+//
+// If there is no groupBy, the function treats all workstatuses as a single
+// group.
+func handleAggregation(scName string, scData *statusCollectorData) *v1alpha1.NamedStatusCombination {
+	// The aggregation requires grouping workstatuses by tuples of groupBy values,
+	// where for N groupBy expressions, a group key would be a tuple of N values.
+	// To achieve this grouping, we maintain two maps:
+	// 1. groupBy Evaluation Value -> unique ID
+	// 2. N-values tuple -> []*workStatusData
+	// 		Where the N-values tuple is the string concatenation of the assigned numbers
+	// 		seperated by commas, ordered by the groupBy expressions in scData.GroupBy slice.
+	generator := 0
+	ValueToNumber := map[any]int{}
+	orderedTupleToWorkStatusGroup := map[string][]*workStatusData{}
+
+	if len(scData.GroupBy) > 0 {
+		for _, wsData := range scData.wecToData {
+			valuesTuple := make([]string, 0, len(scData.GroupBy))
+			for _, groupByNamedExp := range scData.GroupBy {
+				groupByValue := wsData.groupByEval[groupByNamedExp.Name]
+				// ensure unique number mapping for the value
+				uid, exists := ValueToNumber[groupByValue.Value()]
```

Comment:
I am not sure that Golang equality does what is needed here.

Reply:
If Golang equality is the same as the equality needed here, then there is a simpler and more direct approach than using `ValueToNumber` and `orderedTupleToWorkStatusGroup`. Instead it could be something like the following, using a recursive indexed data structure.

```
	grouped := IndexedOrNot{}
	for _, wsData := range scData.wecToData {
		grouped.insert(scData.GroupBy, wsData)
	}
```

where

```
type IndexedOrNot struct {
	indexed map[any]*IndexedOrNot
	not     []*workStatusData
}

func (ion *IndexedOrNot) insert(groupBy []v1alpha1.NamedExpression, wsData *workStatusData) {
	this := ion
	for j := 0; j < len(groupBy); j++ {
		if this.indexed == nil {
			this.indexed = map[any]*IndexedOrNot{}
		}
		groupByVal := wsData.groupByEval[groupBy[j].Name]
		groupByAny := groupByVal.Value()
		next := this.indexed[groupByAny]
		if next == nil {
			next = &IndexedOrNot{}
			this.indexed[groupByAny] = next
		}
		this = next
	}
	this.not = append(this.not, wsData)
}
```